points/forwarder.go

```diff
@@ -0,0 +1,130 @@
+package points
+
+import (
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/rcrowley/go-metrics"
+	"github.com/wavefronthq/go-proxy/api"
+)
+
+// Interface that forwards points to a Wavefront instance.
+type PointForwarder interface {
+	init()
+	addPoint(point string)
+	incrementBlockedPoint()
+	stop()
+}
+
+type DefaultPointForwarder struct {
+	name            string
+	prefix          string
+	workUnitId      string
+	dataFormat      string
+	points          []string
+	maxBufferSize   int
+	maxFlushSize    int
+	mtx             sync.Mutex
+	api             api.WavefrontAPI
+	pushTicker      *time.Ticker
+	pointsReceived  metrics.Counter
+	pointsBlocked   metrics.Counter
+	pointsQueued    metrics.Counter
+	pointsSent      metrics.Counter
+	pointsFlushTime metrics.Timer
+}
+
+func (f *DefaultPointForwarder) init() {
+	f.pointsReceived = metrics.GetOrRegisterCounter("points."+f.prefix+".received", nil)
+	f.pointsBlocked = metrics.GetOrRegisterCounter("points."+f.prefix+".blocked", nil)
+	f.pointsQueued = metrics.GetOrRegisterCounter("points."+f.prefix+".queued", nil)
+	f.pointsSent = metrics.GetOrRegisterCounter("points."+f.prefix+".sent", nil)
+	f.pointsFlushTime = metrics.GetOrRegisterTimer("push."+f.prefix+".duration", nil)
+	go f.flushPoints()
+}
+
+func (f *DefaultPointForwarder) flushPoints() {
+	for range f.pushTicker.C {
+		f.pointsFlushTime.Time(func() {
+			f.post(f.getPointsBatch())
+		})
+	}
+	log.Printf("%s: exiting flushPoints", f.name)
+}
+
+func (f *DefaultPointForwarder) stop() {
+	f.pushTicker.Stop()
+}
+
+func min(x, y int) int {
+	if x < y {
+		return x
+	}
+	return y
+}
+
+func (f *DefaultPointForwarder) getPointsBatch() []string {
+	f.mtx.Lock()
+	currLen := len(f.points)
+	batchSize := min(currLen, f.maxFlushSize)
+	batchPoints := f.points[:batchSize]
+	f.points = f.points[batchSize:currLen]
+	f.mtx.Unlock()
+	return batchPoints
+}
+
+func (f *DefaultPointForwarder) buffer(points []string) {
+	f.mtx.Lock()
```

Comment:
is "defer unlock" idiomatic here? or maybe golang has something similar to python's context guard?

Reply:
Yes, makes sense to do `defer f.mtx.Unlock()` at the top here.